### PR TITLE
Code coverage!

### PR DESCRIPTION
**Pull request by afiune**
## Description
As a Team, we need to have visibility over our code coverage, moving
forward every line of code we add will have to have tests attached to
it, unless there is a strong and compelling reason not to add them.

This change introduces code coverage to this repository, there is a new
studio helper called `code_coverage` that will execute all the tests and
print out both, terminal and HTML based output of the coverage of the
go packages inside this repository.

Here is an example of the output of the studio helper method:
```
[1][default:/src:0]# code_coverage
  hab-studio: Running unit tests
?   	github.com/chef/chef-analyze	[no test files]
?   	github.com/chef/chef-analyze/cmd	[no test files]
ok  	github.com/chef/chef-analyze/pkg/config	0.100s	coverage: 84.2% of statements
ok  	github.com/chef/chef-analyze/pkg/reporting	0.010s	coverage: 18.2% of statements
  hab-studio: Output coverage profile information for each function
github.com/chef/chef-analyze/pkg/config/config.go:55:		ActiveProfile		100.0%
github.com/chef/chef-analyze/pkg/config/config.go:60:		SwitchProfile		85.7%
github.com/chef/chef-analyze/pkg/config/config.go:77:		FromViper		84.6%
github.com/chef/chef-analyze/pkg/config/config.go:109:		NewDefault		100.0%
github.com/chef/chef-analyze/pkg/config/config.go:114:		New			80.0%
github.com/chef/chef-analyze/pkg/config/config.go:147:		FindCredentialsFile	100.0%
github.com/chef/chef-analyze/pkg/config/config.go:161:		FindConfigFile		84.2%
github.com/chef/chef-analyze/pkg/reporting/chef_client.go:14:	NewChefClient		42.9%
github.com/chef/chef-analyze/pkg/reporting/cookbooks.go:12:	Cookbooks		21.4%
github.com/chef/chef-analyze/pkg/reporting/nodes.go:11:		Nodes			0.0%
total:								(statements)		60.0%
  hab-studio: Generate HTML representation of coverage profile
  hab-studio: HTML output written to '/src/coverage.html' (open file with a web browser)
```

When opening the `coverage.html` file locally, you will see the following page on your web browser:
image

## Post Coverage On PRs
Additionally, I have implemented an automation that will upload the generated
coverage output files into the buildkite job and also, post back the results to
the Pull Request as a Comment posted by our CI user (expeditor), this will allow
us to have the numbers upfront and not degrade our code coverage.

Look at the latest automated coverage post/callback: https://github.com/chef/chef-analyze/pull/38#issuecomment-548179546

## Next Steps
- [x] Store the global % of code coverage.
- [x] Create a badge on the main `README.md`
- [x] Compare the % between master and open PRs.

Signed-off-by: Salim Afiune <afiune@chef.io>

## Related Issue

Not related directly but I want to link them since this was a conversation we
had in standup about testing as tech depot: https://github.com/chef/chef-dk/issues/2135

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

**Comment by afiune** [comment]
Code Coverage: github.com/chef/chef-analyze/pkg/config/config.go:55: ActiveProfile 100.0% github.com/chef/chef-analyze/pkg/config/config.go:60: SwitchProfile 85.7% github.com/chef/chef-analyze/pkg/config/config.go:77: FromViper 84.6% github.com/chef/chef-analyze/pkg/config/config.go:109: NewDefault 100.0% github.com/chef/chef-analyze/pkg/config/config.go:114: New 80.0% github.com/chef/chef-analyze/pkg/config/config.go:147: FindCredentialsFile 100.0% github.com/chef/chef-analyze/pkg/config/config.go:161: FindConfigFile 84.2% github.com/chef/chef-analyze/pkg/reporting/chef_client.go:14: NewChefClient 42.9% github.com/chef/chef-analyze/pkg/reporting/cookbooks.go:12: Cookbooks 21.4% github.com/chef/chef-analyze/pkg/reporting/nodes.go:11: Nodes 0.0% total: (statements) 60.0%

**Comment by afiune** [comment]
# Code Coverage 

Code details | Function name | % Coverage 
-----------------------------------------

**Comment by afiune** [comment]
# Code Coverage 

Code details | Function name | % Coverage 
------------ | ------------- | ---------- 
 github.com/chef/chef-analyze/pkg/config/config.go:55: | ActiveProfile | 100.0%
github.com/chef/chef-analyze/pkg/config/config.go:60: | SwitchProfile | 85.7%
github.com/chef/chef-analyze/pkg/config/config.go:77: | FromViper | 84.6%
github.com/chef/chef-analyze/pkg/config/config.go:109: | NewDefault | 100.0%
github.com/chef/chef-analyze/pkg/config/config.go:114: | New | 80.0%
github.com/chef/chef-analyze/pkg/config/config.go:147: | FindCredentialsFile | 100.0%
github.com/chef/chef-analyze/pkg/config/config.go:161: | FindConfigFile | 84.2%
github.com/chef/chef-analyze/pkg/reporting/chef_client.go:14: | NewChefClient | 42.9%
github.com/chef/chef-analyze/pkg/reporting/cookbooks.go:12: | Cookbooks | 21.4%
github.com/chef/chef-analyze/pkg/reporting/nodes.go:11: | Nodes | 0.0%
total: | (statements) | 60.0%

File: .expeditor/update_coverage.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+export COVERAGE_DATE
+COVERAGE_DATE="$(date +"%Y-%m-%dT%H:%M:%S")"
+
+# TODO @afiune How can we download artifacts from previous buildkite builds?
+# use instead the agent
+#buildkite-agent artifact download "coverage/*" --build XXX-XXX-XXX(?)
+hab studio run "source .studiorc && code_coverage"
+
+aws --profile chef-cd s3 sync coverage "s3://chef-workstation-coverage/$BUILDKITE_PIPELINE_SLUG/current"
```
Comment (tyler-ball) [inline .expeditor/update_coverage.sh]: What is an example of `BUILDKITE_PIPELINE_SLUG`? I want to make sure we are able to differentiate between the code coverage of `chef-analyze` and some (theoretical) other project that uploads coverage

Comment (afiune) [reply .expeditor/update_coverage.sh]: `BUILDKITE_PIPELINE_SLUG="chef-chef-analyze-master-verify"`

You can see all the environment variables on the last tab of each buildkite job:
image

https://buildkite.com/chef/chef-chef-analyze-master-verify/builds/127#c9575ee5-0d46-4b9b-ada5-51b31007013c

**Pull request merged**